renormalizer/vibration/vscf.py

```diff
@@ -0,0 +1,119 @@
+
+import numpy as np
+import scipy
+import opt_einsum as oe
+import logging
+
+from renormalizer.mps.backend import xp, OE_BACKEND
+from renormalizer.mps.lib import Environ, cvec2cmat
+from renormalizer.mps import Mpo, Mps
+from renormalizer.mps.matrix import asnumpy, asxp
+
+logger = logging.getLogger(__name__)
+
+class Vscf():
+    r"""
+        Vibrational Self-Consistent Field 
+        
+        The algorithm is like a DMRG sweep.
+    """
+
+
+    def __init__(self, model, mps=None):
+        self.model = model
+        self.c = [None,]*model.nsite # coefficient
+        self.e = [None,]*model.nsite # energy
+
+        if "h_mpo" in model.mpos.keys():
+            logger.info("load h_mpo form model.mpos")
+            self.h_mpo = model.mpos["h_mpo"]
+        else:
+            self.h_mpo = Mpo(model)
+        if mps is None:
+            self.mps = Mps.hartree_product_state(self.model, dict())
+        else:
+            self.mps = mps
+    
+    def kernel(self, nsweeps=100):
```

Comment:
A lot of the code is in duplication with `gs.py`. Is it possible to call `optimize_mps` in this function?

Reply:
It is not convenient to call `optimize_mps` unless `optimize_mps` is adapted to vscf,  including the convergence criteria, how to solve the eigenproblem, etc.
Considering vscf is rarely used and the current sweep-like implementation is an ad-hoc method, I think it is better to separate it from the more complex DMRG code.

Reply:
OK. The primary differences are that 1) for the eigenproblem, all roots are solved and stored, but the MPS is only updated by the lowest eigenstate and 2) for the convergence criteria, both energy and coefficient are taken into account. Am I right?
If it is really "ad-hoc" then in the future we should replaced it by another standard implementation. Otherwise maintaining two pieces of similar code would be troublesome.

Reply:
For vscf, the standard implementation is to update the coeff of each site together like HF.

Reply:
Yes, you are right.